packages/grafana-ui/src/components/BarChart/BarChart.tsx

```diff
@@ -0,0 +1,257 @@
+// Library
+import { getConfig } from './bars';
+
+import React, { useCallback, useMemo, useRef } from 'react';
+import {
+  compareDataFrameStructures,
+  DataFrame,
+  DefaultTimeZone,
+  formattedValueToString,
+  getFieldDisplayName,
+  getFieldSeriesColor,
+  TimeRange,
+} from '@grafana/data';
+
+import { VizLayout } from '../VizLayout/VizLayout';
+
+// Types
+import { VizOrientation } from '@grafana/data';
+import { Themeable } from '../../types';
+import { BarChartOptions, BarValueVisibility } from './types';
+import { useRevision } from '../uPlot/hooks';
+import { UPlotChart } from '../uPlot/Plot';
+import { UPlotConfigBuilder } from '../uPlot/config/UPlotConfigBuilder';
+import { AxisPlacement, ScaleDistribution } from '../uPlot/config';
+import { useTheme } from '../../themes';
+import { GraphNGLegendEvent, GraphNGLegendEventMode } from '../GraphNG/types';
+import { LegendDisplayMode, VizLegendItem } from '../VizLegend/types';
+import { VizLegend } from '../VizLegend/VizLegend';
+
+export interface Props extends Themeable, BarChartOptions {
+  height: number;
+  width: number;
+  data: DataFrame;
+  onLegendClick?: (event: GraphNGLegendEvent) => void;
+  onSeriesColorChange?: (label: string, color: string) => void;
+}
+
+/**
+ * @alpha
+ */
+export const BarChart: React.FunctionComponent<Props> = ({
+  width,
+  height,
+  data,
+  orientation,
+  groupWidth,
+  barWidth,
+  showValue,
+  legend,
+  onLegendClick,
+  onSeriesColorChange,
+  ...plotProps
+}) => {
+  if (!data || data.fields.length < 2) {
+    return <div>Missing data</div>;
+  }
+
+  // dominik? TODO? can this all be moved into `useRevision`
+  const compareFrames = useCallback((a?: DataFrame | null, b?: DataFrame | null) => {
+    if (a && b) {
+      return compareDataFrameStructures(a, b);
+    }
+    return false;
+  }, []);
+
+  const configRev = useRevision(data, compareFrames);
+
+  const theme = useTheme();
+
+  // Updates only when the structure changes
+  const configBuilder = useMemo(() => {
+    if (!orientation || orientation === VizOrientation.Auto) {
+      orientation = width < height ? VizOrientation.Horizontal : VizOrientation.Vertical;
+    }
+
+    // bar orientation -> x scale orientation & direction
+    const xOri = orientation === VizOrientation.Horizontal ? 1 : 0;
+    const xDir = orientation === VizOrientation.Horizontal ? -1 : 1;
+
+    const formatValue =
+      showValue !== BarValueVisibility.Never
+        ? (seriesIdx: number, value: any) => formattedValueToString(data.fields[seriesIdx].display!(value))
+        : undefined;
+
+    const config = getConfig(xOri, xDir, groupWidth, barWidth, formatValue);
```

Comment:
probably should accept an options object instead of continuing to grow the args list.